service/dynamodb/dynamodbattribute/encode.go

```diff
@@ -0,0 +1,465 @@
+			elemFn = func(elem dynamodb.AttributeValue) error {
+				if elem.S == nil {
+					return &InvalidMarshalError{msg: "string set must only contain non-nil strings"}
+				}
+				av.SS = append(av.SS, elem.S)
+				return nil
+			}
+		} else { // List
+			av.L = make([]*dynamodb.AttributeValue, 0, v.Len())
+			elemFn = func(elem dynamodb.AttributeValue) error {
+				av.L = append(av.L, &elem)
+				return nil
+			}
+		}
+
+		if n, err := e.encodeList(v, fieldTag, elemFn); err != nil {
+			return err
+		} else if n == 0 {
+			encodeNull(av)
+		}
+	}
+
+	return nil
+}
+
+func (e *Encoder) encodeList(v reflect.Value, fieldTag tag, elemFn func(dynamodb.AttributeValue) error) (int, error) {
+	count := 0
+	for i := 0; i < v.Len(); i++ {
+		elem := dynamodb.AttributeValue{}
+		err := e.encode(&elem, v.Index(i), tag{OmitEmpty: fieldTag.OmitEmptyElem})
+		skip, err := keepOrOmitEmpty(fieldTag.OmitEmptyElem, &elem, err)
+		if err != nil {
+			return 0, err
+		} else if skip {
+			continue
+		}
+
+		if err := elemFn(elem); err != nil {
+			return 0, err
+		}
+		count++
+	}
+
+	return count, nil
+}
+
+func (e *Encoder) encodeScalar(av *dynamodb.AttributeValue, v reflect.Value, fieldTag tag) error {
+	switch typed := v.Interface().(type) {
+	case bool:
+		av.BOOL = new(bool)
+		*av.BOOL = typed
+	case string:
+		if err := e.encodeString(av, v); err != nil {
+			return err
+		}
+	default:
+		// Fallback to encoding numbers, will return invalid type if not supported
+		if err := e.encodeNumber(av, v); err != nil {
+			return err
+		}
+		if fieldTag.AsString && av.NULL == nil && av.N != nil {
+			av.S = av.N
+			av.N = nil
+		}
+	}
+
+	return nil
+}
+
+func (e *Encoder) encodeNumber(av *dynamodb.AttributeValue, v reflect.Value) error {
+	if used, err := tryMarshaler(av, v); used {
+		return err
+	}
+
+	var out string
+	switch typed := v.Interface().(type) {
+	case int:
+		out = encodeInt(int64(typed))
+	case int8:
+		out = encodeInt(int64(typed))
+	case int16:
+		out = encodeInt(int64(typed))
+	case int32:
+		out = encodeInt(int64(typed))
+	case int64:
+		out = encodeInt(typed)
+	case uint:
+		out = encodeUint(uint64(typed))
+	case uint8:
+		out = encodeUint(uint64(typed))
+	case uint16:
+		out = encodeUint(uint64(typed))
+	case uint32:
+		out = encodeUint(uint64(typed))
+	case uint64:
+		out = encodeUint(typed)
+	case float32:
+		out = encodeFloat(float64(typed))
+	case float64:
+		out = encodeFloat(typed)
+	default:
+		return &unsupportedMarshalTypeError{Type: v.Type()}
+	}
+
+	av.N = &out
+
+	return nil
+}
+
+func (e *Encoder) encodeString(av *dynamodb.AttributeValue, v reflect.Value) error {
+	if used, err := tryMarshaler(av, v); used {
+		return err
+	}
+
+	switch typed := v.Interface().(type) {
+	case string:
+		if len(typed) == 0 && e.NullEmptyString {
+			encodeNull(av)
+		} else {
+			av.S = &typed
+		}
+	default:
+		return &unsupportedMarshalTypeError{Type: v.Type()}
+	}
+
+	return nil
+}
+
+func encodeInt(i int64) string {
```

Comment:
I'll try that, and see if I can use the Sprintf, the benefit of the `strconv.Format` methods are they don't need to do the extra processing Sprintf performs. Minor optimization in this case since we know the types.